File: core/test/TestFixtures/Points.hs
module TestFixtures.Points where
import Test.Hspec
import TestFixtures.GameState

import Domain.GameState
import Domain.Structure
import Domain.Player
import Domain.Resource

import Core.Points

import Data.Map (Map, fromList, adjust)

testPoints =
    describe "Points.calculatePoints" $
        it "with 3 players" $
            calculatePoints gameStateForPoints `shouldBe` pointsExpected

gameStateForPoints =
    defaultGameState {
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                coins = 0,
                battleTokens = [-1, -1, -1, -1, -1, -1],
                pointActions = [
                    return (adjust (3+) WonderP),
                    return (adjust (6+) CivilianP)
                ],
                scientificSymbols = fromList [
                    (Compass, 1),
                    (Gears, 1),
                    (Tablet, 1)
                ],
                scientificActions = []
            }),
            ("b", (defaultPlayerState "b") {
                coins = 9,
                battleTokens = [0, 1, -1, 3, 5, 5],
                pointActions = [],
                scientificSymbols = fromList [
                    (Compass, 4),
                    (Gears, 4),
                    (Tablet, 4)
                ],
                scientificActions = [
                    return (adjust (1+) Compass),
                    return (adjust (1+) Gears)
                ]
            }),
            ("c", (defaultPlayerState "c") {
                coins = 4,
                battleTokens = [1, 1, 3, 3, 5, 5],
                pointActions = [
                    return (adjust (10+) GuildsP),
                    return (adjust (10+) GuildsP)
                ],
                scientificSymbols = fromList [(Compass, 0), (Gears, 0), (Tablet, 0)],
                scientificActions = [
                    return (adjust (1+) Compass)
                ]
            })
        ]
    }

pointsExpected :: Map Domain.Player.Name (Map PointCategory Int)
pointsExpected = 
    fromList [
        ("a", fromList [(MilitaryP, -6), (TreasuryP, 0), (WonderP, 3), (CivilianP, 6), (ScientificP, 10), (CommercialP, 0), (GuildsP, 0)]),
        ("b", fromList [(MilitaryP, 13), (TreasuryP, 3), (WonderP, 0), (CivilianP, 0), (ScientificP, 94), (CommercialP, 0), (GuildsP, 0)]),
        ("c", fromList [(MilitaryP, 18), (TreasuryP, 1), (WonderP, 0), (CivilianP, 0), (ScientificP, 1), (CommercialP, 0), (GuildsP, 20)])
    ]

File: core/src/tests/points.rs
use super::helpers::default_player_state;
use crate::domain::{
    GameState, PlayerState,
    PointCategory::{CivilianP, CommercialP, GuildP, MilitaryP, ScientificP, TreasuryP, WonderP},
    PointsMap,
    ScientificSymbol::{Compass, Gears, Tablet},
    ScientificSymbolCounts, ScientificSymbolsProduced,
};
use maplit::hashmap;

#[test]
fn test_calculate_points() {
    assert_eq!(
        GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState{
                    coins: 0,
                    battle_tokens: vec![-1, -1, -1, -1, -1, -1],
                    point_actions: vec![
                        Box::new(|_:&_, points_map: &mut PointsMap| {
                            points_map.add(&WonderP, 3)
                        }),
                        Box::new(|_:&_, points_map: &mut PointsMap| {
                            points_map.add(&CivilianP, 6)
                        }),
                    ],
                    scientific_symbols_produced: ScientificSymbolsProduced{
                        compass: 1,
                        gears: 1,
                        tablet: 1,
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "b".to_string() => PlayerState{
                    coins: 9,
                    battle_tokens: vec![0, 1, -1, 3, 5, 5],
                    scientific_symbols_produced: ScientificSymbolsProduced{
                        compass: 4,
                        gears: 4,
                        tablet: 4,
                        actions: vec![
                            Box::new(|_, symbol_counts: &mut ScientificSymbolCounts| {
                                symbol_counts[1] += 1;
                                symbol_counts[2] += 1;
                            })
                        ],
                        any_symbols: vec![&[Tablet, Gears, Compass]],
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "c".to_string() => PlayerState{
                    coins: 4,
                    point_actions: vec![
                        Box::new(|_:&_, points_map: &mut PointsMap| {
                            points_map.add(&GuildP, 10);
                        }),
                        Box::new(|_:&_, points_map: &mut PointsMap| {
                            points_map.add(&GuildP, 10);
                        }),
                    ],
                    ..default_player_state()
                }
            },
            ..Default::default()
        }
        .calculate_points(),
        hashmap! {
            "a".to_string() => hashmap!{
                MilitaryP => -6,
                TreasuryP => 0,
                WonderP => 3,
                CivilianP => 6,
                ScientificP => 10,
                CommercialP => 0,
                GuildP => 0,
            },
            "b".to_string() => hashmap!{
                MilitaryP => 13,
                TreasuryP => 3,
                WonderP => 0,
                CivilianP => 0,
                ScientificP => 110,
                CommercialP => 0,
                GuildP => 0,
            },
            "c".to_string() => hashmap!{
                MilitaryP => 0,
                TreasuryP => 1,
                WonderP => 0,
                CivilianP => 0,
                ScientificP => 0,
                CommercialP => 0,
                GuildP => 20,
            },
        }
    )
}
